File: Anime4K-Upscaler/Models/CompressModels.swift
// Anime4K-Upscaler/Models/CompressModels.swift
// Domain models for the Compress feature — encoder, content type, HDR, configuration, job, argument builder.

import Foundation
import SwiftUI
import Observation

// MARK: - Compress Encoder

/// Encoder options for the Compress feature.
enum CompressEncoder: String, CaseIterable, Identifiable, Sendable {
    case hevcVideoToolbox = "hevc_videotoolbox"
    case svtAV1           = "libsvtav1"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .hevcVideoToolbox: return "HEVC (Hardware)"
        case .svtAV1:           return "SVT-AV1 (Software)"
        }
    }

    var subtitle: String {
        switch self {
        case .hevcVideoToolbox: return "VideoToolbox GPU — Fastest encoding"
        case .svtAV1:           return "SVT-AV1 CPU — Maximum storage saving"
        }
    }

    var symbolName: String {
        switch self {
        case .hevcVideoToolbox: return "bolt.fill"
        case .svtAV1:           return "cpu"
        }
    }

    /// Default quality value for each encoder.
    var defaultQuality: Int {
        switch self {
        case .hevcVideoToolbox: return 68
        case .svtAV1:           return 24
        }
    }

    /// Label describing the quality scale.
    var qualityLabel: String {
        switch self {
        case .hevcVideoToolbox: return "Quality (0–100, Higher = Better)"
        case .svtAV1:           return "CRF (0–63, Lower = Better)"
        }
    }

    /// Maximum valid quality value.
    var maxQuality: Int {
        switch self {
        case .hevcVideoToolbox: return 100
        case .svtAV1:           return 63
        }
    }

    var usesCRF: Bool { self == .svtAV1 }
}

// MARK: - Content Type

/// Content type selection for tuning anime vs live-action encoding.
enum ContentType: String, CaseIterable, Identifiable, Sendable {
    case liveAction = "live_action"
    case anime      = "anime"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .liveAction: return "Live Action"
        case .anime:      return "Anime / Animation"
        }
    }

    var subtitle: String {
        switch self {
        case .liveAction: return "Standard live-action video encoding"
        case .anime:      return "Enables B-Frames & Long GOP options"
        }
    }

    var symbolName: String {
        switch self {
        case .liveAction: return "video"
        case .anime:      return "sparkles.tv"
        }
    }
}

// MARK: - HDR Mode (auto-detected per file)

/// HDR status detected from ffprobe color_transfer metadata.
enum HDRMode: String, Sendable {
    case sdr  = "SDR"
    case hdr10 = "HDR10"

    var displayName: String {
        switch self {
        case .sdr:  return "SDR (Rec.709)"
        case .hdr10: return "HDR10 (Pass-through)"
        }
    }
}

// MARK: - Compress Configuration

/// User-configurable settings for the Compress feature.
struct CompressConfiguration: Sendable {
    var encoder: CompressEncoder = .hevcVideoToolbox
    var quality: Int = 68
    var contentType: ContentType = .liveAction
    var bFrames: Int = 3
    var longGOPEnabled: Bool = false

    static let `default` = CompressConfiguration()
}

// MARK: - Compress Job

/// A single file being compressed — tracks state, progress, and log output.
@MainActor @Observable
final class CompressJob: Identifiable {
    let id: UUID
    let file: VideoFile
    let configuration: CompressConfiguration

    var state: JobState = .idle
    var progress: Double = 0.0
    var currentFrame: Int = 0
    var currentTime: String = "00:00:00.00"
    var speed: String = "0.0x"
    var fps: String = "0"
    var outputURL: URL?
    var logLines: [String] = []
    var errorMessage: String?
    var startDate: Date?
    var endDate: Date?
    var hdrMode: HDRMode = .sdr

    @ObservationIgnored var processHandle: Process?

    init(file: VideoFile, configuration: CompressConfiguration, outputDirectory: URL? = nil) {
        self.id = UUID()
        self.file = file
        self.configuration = configuration

        let baseDir = outputDirectory ?? file.url.deletingLastPathComponent()
        let outName = "\(file.fileName)_compressed.\(file.fileExtension)"
        self.outputURL = baseDir.appendingPathComponent(outName)
    }

    var elapsedTime: TimeInterval? {
        guard let start = startDate else { return nil }
        return (endDate ?? Date()).timeIntervalSince(start)
    }

    var formattedElapsedTime: String? {
        guard let elapsed = elapsedTime else { return nil }
        let hours = Int(elapsed) / 3600
        let minutes = (Int(elapsed) % 3600) / 60
        let seconds = Int(elapsed) % 60
        if hours > 0 {
            return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%02d:%02d", minutes, seconds)
        }
    }

    /// Append a log line (capped at 500 lines to prevent memory bloat).
    /// Uses in-place `removeSubrange` instead of allocating a new Array.
    func appendLog(_ line: String) {
        logLines.append(line)
        let overflow = logLines.count - 500
        if overflow > 0 {
            logLines.removeSubrange(0..<overflow)
        }
    }
}

// MARK: - Compress Argument Builder

/// Assembles the complete FFmpeg argument array for a compress job.
struct CompressArgumentBuilder {

    static func build(
        inputURL: URL,
        outputURL: URL,
        configuration: CompressConfiguration,
        hdrMode: HDRMode
    ) -> [String] {
        var args: [String] = []
        args.reserveCapacity(32) // ~25-30 args typical; avoids backing-store reallocations

        // Global flags
        args.append(contentsOf: ["-hide_banner", "-v", "error", "-stats"])
        args.append(contentsOf: ["-y"])
        args.append(contentsOf: ["-i", inputURL.path])

        // Map: first video, all audio, all subtitles
        args.append(contentsOf: ["-map", "0:v:0", "-map", "0:a?", "-map", "0:s?"])

        // Copy audio and subtitles
        args.append(contentsOf: ["-c:a", "copy", "-c:s", "copy"])

        // Encoder-specific arguments
        switch configuration.encoder {
        case .svtAV1:
            var svtParams = "tune=0:"
            if hdrMode == .hdr10 {
                svtParams += "enable-hdr=1:"
            } else {
                svtParams += "enable-hdr=0:color-primaries=1:transfer-characteristics=1:matrix-coefficients=1:range=1:"
            }
            svtParams = String(svtParams.dropLast()) // Remove trailing colon

            args.append(contentsOf: [
                "-c:v", "libsvtav1",
                "-preset", "4",
                "-crf", "\(configuration.quality)",
                "-pix_fmt", "yuv420p10le",
                "-svtav1-params", svtParams
            ])

        case .hevcVideoToolbox:
            args.append(contentsOf: [
                "-c:v", "hevc_videotoolbox",
                "-q:v", "\(configuration.quality)",
                "-pix_fmt", "yuv420p10le"
            ])

            // Color tags based on HDR detection
            if hdrMode == .hdr10 {
                args.append(contentsOf: [
                    "-color_primaries", "bt2020",
                    "-color_trc", "smpte2084",
                    "-colorspace", "bt2020nc"
                ])
            } else {
                args.append(contentsOf: [
                    "-color_primaries", "bt709",
                    "-color_trc", "bt709",
                    "-colorspace", "bt709",
                    "-color_range", "tv"
                ])
            }
        }

        // Anime-specific arguments
        if configuration.contentType == .anime {
            if configuration.bFrames > 0 {
                args.append(contentsOf: ["-bf", "\(configuration.bFrames)"])
            }
            if configuration.longGOPEnabled {
                args.append(contentsOf: [
                    "-g", "240",
                    "-keyint_min", "240",
                    "-sc_threshold", "0"
                ])
            }
        }

        // Output file
        args.append(outputURL.path)

        return args
    }
}

File: windows/src/models/CompressModels.cpp
#include "CompressModels.h"
#include "../utils/HardwareDetector.h"
#include <QFileInfo>
#include <QDir>

// MARK: - Compress Encoder Helper Functions

QString displayName(CompressEncoder encoder) {
    switch (encoder) {
        case CompressEncoder::HEVC_NVENC: return "HEVC (NVIDIA RTX)";
        case CompressEncoder::HEVC_AMF:   return "HEVC (AMD)";
        case CompressEncoder::HEVC_QSV:   return "HEVC (Intel)";
        case CompressEncoder::SVT_AV1:     return "SVT-AV1 (Software)";
    }
    return QString();
}

QString subtitle(CompressEncoder encoder) {
    DeviceHardwareProfile profile = DeviceHardwareProfile::detect();
    switch (encoder) {
        case CompressEncoder::HEVC_NVENC: return QString("NVIDIA NVENC on %1 \u2014 Fastest encoding").arg(profile.gpuName);
        case CompressEncoder::HEVC_AMF:   return QString("AMD AMF on %1 \u2014 Fastest encoding").arg(profile.gpuName);
        case CompressEncoder::HEVC_QSV:   return QString("Intel QSV on %1 \u2014 Fastest encoding").arg(profile.gpuName);
        case CompressEncoder::SVT_AV1:     return QString("SVT-AV1 on %1-core CPU \u2014 Maximum storage saving").arg(profile.cpuCoreCount);
    }
    return QString();
}

QString symbolName(CompressEncoder encoder) {
    switch (encoder) {
        case CompressEncoder::HEVC_NVENC:
        case CompressEncoder::HEVC_AMF:
        case CompressEncoder::HEVC_QSV:
            return "bolt.fill";
        case CompressEncoder::SVT_AV1:
            return "cpu";
    }
    return QString();
}

int defaultQuality(CompressEncoder encoder) {
    switch (encoder) {
        case CompressEncoder::HEVC_NVENC:
        case CompressEncoder::HEVC_AMF:
        case CompressEncoder::HEVC_QSV:
            return 68;
        case CompressEncoder::SVT_AV1:
            return 24;
    }
    return 68;
}

QString qualityLabel(CompressEncoder encoder) {
    switch (encoder) {
        case CompressEncoder::HEVC_NVENC:
        case CompressEncoder::HEVC_AMF:
        case CompressEncoder::HEVC_QSV:
            return "Quality (0\u2013100, Higher = Better)";
        case CompressEncoder::SVT_AV1:
            return "CRF (0\u201363, Lower = Better)";
    }
    return QString();
}

int maxQuality(CompressEncoder encoder) {
    switch (encoder) {
        case CompressEncoder::HEVC_NVENC:
        case CompressEncoder::HEVC_AMF:
        case CompressEncoder::HEVC_QSV:
            return 100;
        case CompressEncoder::SVT_AV1:
            return 63;
    }
    return 100;
}

bool usesCRF(CompressEncoder encoder) {
    return encoder == CompressEncoder::SVT_AV1;
}

// MARK: - Content Type Helper Functions

QString displayName(ContentType type) {
    switch (type) {
        case ContentType::LiveAction: return "Live Action";
        case ContentType::Anime:      return "Anime / Animation";
    }
    return QString();
}

QString subtitle(ContentType type) {
    switch (type) {
        case ContentType::LiveAction: return "Standard live-action video encoding";
        case ContentType::Anime:      return "Enables B-Frames & Long GOP options";
    }
    return QString();
}

QString symbolName(ContentType type) {
    switch (type) {
        case ContentType::LiveAction: return "video";
        case ContentType::Anime:      return "sparkles.tv";
    }
    return QString();
}

// MARK: - HDR Mode Helper Functions

QString displayName(HDRMode mode) {
    switch (mode) {
        case HDRMode::SDR:   return "SDR (Rec.709)";
        case HDRMode::HDR10: return "HDR10 (Pass-through)";
    }
    return QString();
}

// MARK: - Compress Configuration

CompressConfiguration CompressConfiguration::defaultConfig() {
    return CompressConfiguration();
}

// MARK: - Compress Job

CompressJob::CompressJob(const VideoFile& f, const CompressConfiguration& c, const QString& outputDirectory, QObject* parent)
    : QObject(parent), file(f), configuration(c) {
    id = QUuid::createUuid();
    QString baseDir = outputDirectory.isEmpty() ? QFileInfo(file.filePath).absolutePath() : outputDirectory;
    QString outName = QString("%1_compressed.%2").arg(file.fileName).arg(file.fileExtension);
    outputPath = QDir(baseDir).filePath(outName);
}

double CompressJob::elapsedTime() const {
    if (!startDate.isValid()) return 0.0;
    QDateTime end = endDate.isValid() ? endDate : QDateTime::currentDateTime();
    return startDate.msecsTo(end) / 1000.0;
}

QString CompressJob::formattedElapsedTime() const {
    double elapsed = elapsedTime();
    int hours = static_cast<int>(elapsed) / 3600;
    int minutes = (static_cast<int>(elapsed) % 3600) / 60;
    int seconds = static_cast<int>(elapsed) % 60;
    if (hours > 0) {
        return QString("%1:%2:%3")
            .arg(hours, 2, 10, QChar('0'))
            .arg(minutes, 2, 10, QChar('0'))
            .arg(seconds, 2, 10, QChar('0'));
    } else {
        return QString("%1:%2")
            .arg(minutes, 2, 10, QChar('0'))
            .arg(seconds, 2, 10, QChar('0'));
    }
}

void CompressJob::appendLog(const QString& line) {
    logLines.append(line);
    int overflow = logLines.count() - 500;
    if (overflow > 0) {
        logLines.erase(logLines.begin(), logLines.begin() + overflow);
    }
}

// MARK: - Compress Argument Builder

QStringList CompressArgumentBuilder::build(
    const QString& inputPath,
    const QString& outputPath,
    const CompressConfiguration& configuration,
    HDRMode hdrMode
) {
    QStringList args;
    args.append({"-hide_banner", "-v", "error", "-stats"});
    args.append("-y");
    args.append({"-i", inputPath});

    // Stream maps
    args.append({"-map", "0:v:0", "-map", "0:a?", "-map", "0:s?"});

    // Audio & Subs copy
    args.append({"-c:a", "copy", "-c:s", "copy"});

    // Encoder
    switch (configuration.encoder) {
        case CompressEncoder::SVT_AV1: {
            QString svtParams = "tune=0:";
            if (hdrMode == HDRMode::HDR10) {
                svtParams += "enable-hdr=1:";
            } else {
                svtParams += "enable-hdr=0:color-primaries=1:transfer-characteristics=1:matrix-coefficients=1:range=1:";
            }
            svtParams.chop(1); // remove trailing ':'

            args.append({"-c:v", "libsvtav1",
                         "-preset", "4",
                         "-crf", QString::number(configuration.quality),
                         "-pix_fmt", "yuv420p10le",
                         "-svtav1-params", svtParams});
            break;
        }

        case CompressEncoder::HEVC_NVENC:
            args.append({"-c:v", "hevc_nvenc",
                         "-preset", "p4",
                         "-tune", "hq",
                         "-rc", "vbr",
                         "-cq", QString::number(configuration.quality),
                         "-b:v", "0",
                         "-pix_fmt", "p010le"});
            if (hdrMode == HDRMode::HDR10) {
                args.append({"-color_primaries", "bt2020",
                             "-color_trc", "smpte2084",
                             "-colorspace", "bt2020nc"});
            } else {
                args.append({"-color_primaries", "bt709",
                             "-color_trc", "bt709",
                             "-colorspace", "bt709",
                             "-color_range", "tv"});
            }
            break;

        case CompressEncoder::HEVC_AMF:
            args.append({"-c:v", "hevc_amf",
                         "-quality", "quality",
                         "-rc", "cqp",
                         "-qp_i", QString::number(configuration.quality),
                         "-qp_p", QString::number(configuration.quality),
                         "-pix_fmt", "p010le"});
            if (hdrMode == HDRMode::HDR10) {
                args.append({"-color_primaries", "bt2020",
                             "-color_trc", "smpte2084",
                             "-colorspace", "bt2020nc"});
            } else {
                args.append({"-color_primaries", "bt709",
                             "-color_trc", "bt709",
                             "-colorspace", "bt709",
                             "-color_range", "tv"});
            }
            break;

        case CompressEncoder::HEVC_QSV:
            args.append({"-c:v", "hevc_qsv",
                         "-preset", "medium",
                         "-global_quality", QString::number(configuration.quality),
                         "-pix_fmt", "p010le"});
            if (hdrMode == HDRMode::HDR10) {
                args.append({"-color_primaries", "bt2020",
                             "-color_trc", "smpte2084",
                             "-colorspace", "bt2020nc"});
            } else {
                args.append({"-color_primaries", "bt709",
                             "-color_trc", "bt709",
                             "-colorspace", "bt709",
                             "-color_range", "tv"});
            }
            break;
    }

    // Anime-specific settings
    if (configuration.contentType == ContentType::Anime) {
        if (configuration.bFrames > 0) {
            args.append({"-bf", QString::number(configuration.bFrames)});
        }
        if (configuration.longGOPEnabled) {
            args.append({"-g", "240",
                         "-keyint_min", "240",
                         "-sc_threshold", "0"});
        }
    }

    args.append(outputPath);
    return args;
}
